modules/apps/document-library/document-library-preview-audio/src/main/resources/META-INF/resources/preview/view.jsp

```diff
@@ -23,7 +23,7 @@ List<String> previewFileURLs = (List<String>)request.getAttribute(DLPreviewAudio
 <liferay-util:html-top
 	outputKey="document_library_preview_audio_css"
 >
-	<link href="<%= PortalUtil.getStaticResourceURL(request, application.getContextPath() + "/preview/css/main.css") %>" rel="stylesheet" type="text/css" />
```

Comment:
I have doubts about this, because the dependency is not explicit anywhere, right? You could add it with a TopHeadDynamicInclude or add the dependency in the gradle file if not loading this CSS would make the module unusable. But @izaera knows best.

Reply:
Maybe we can simply use `<aui:script require>` to request the `@liferay/document-library-preview-css` module... Have you tried that? :thinking:

Reply:
If not, it can be added with a `Liferay-JS-Resources-Top-Head` header in the `bnd.bnd` or a TopHeadDynamicInclude class if it needs any activation logic.

Reply:
We use this <html-top> pattern in many places, and URL is not built consistently. In most of the cases where it's used to link to CSS, the CSS it lives in the same bundle as the JSP. But there are cases where css is provided by a separate bundle:
* App Builder: [view.jsp](https://github.com/liferay/liferay-portal/blob/eec91266f7c07a45f5e71106c1acf1f09c48ceff/modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/view.jsp#L20) and [view_apps.jsp](https://github.com/liferay/liferay-portal/blob/eec91266f7c07a45f5e71106c1acf1f09c48ceff/modules/apps/app-builder/app-builder-web/src/main/resources/META-INF/resources/view_apps.jsp#L20) depend on data-engine-taglib 
* Data Engine Taglib: [start.jsp](https://github.com/liferay/liferay-portal/blob/59e8c0d402c6c4b0800e2100affe9d9b854bddaa/modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/start.jsp#L20-L21) depends on css defined by dybamic-data-mapping-form-builder and data-engine-js-components-web
* Journal Web: [import_translation.jsp](https://github.com/liferay/liferay-portal/blob/8009773bf9f0008aee1f7953407574208c14bbd6/modules/apps/journal/journal-web/src/main/resources/META-INF/resources/import_translation.jsp#L36) depend on translation-web

Not many cases, but the need for a mechanism to declare these at JSP level is clear to me if we want to start extratcting common css. An option is to do it at the osgi level, but I'm not sure our loader can do much in that case. The aui:script option sounds reasonable, as long as the common css is declared as in this PR.

Reply:
Maybe a dedicated JSP tag... (just brainstorming)

Reply:
I think I'm not following you guys 😅

Reply:
> Maybe a dedicated JSP tag... (just brainstorming)

Yeah, I was thinking of something like this the other day. <html-top> seems very generic for the purpose of denoting a dependency. Although the functionality may be useful to ensure CSS is available before it's used, it does not help in things like:
* It forces user to create the full resource URL, which is not done consistently across the board (. A module name + local path schema would be a simpler way to express the dependency
* It makes it hard to inline contents: in some scenarios we may prefer to inline CSS rather than adding a <link>. Tag would handle this graciously if the API is designed in terms of a provider module + a relative path plus an `inline` flag
* It's not expressing any explicit dependency we can handle. Still, I don't see how a tag can enforce anything at compile time. I'm not 100% sure if we want/need such a level of compile-time awareness, but it feels a bit off that module A relies on CSS produced by module B, with no compile-time dependency. 

The `Liferay-JS-Resources-Top-Head` (or should it be `Liferay-CSS-Resources-Top-Head`?) would output the CSS in all page requests as long as the module providing them is active right?

Reply:
@dsanz @izaera @nhpatt could we move the discussion about this particular point to slack (or somewhere else) so we can push forward this pr? Brasil is waiting for it :)

Reply:
Commerce also has asked us for guidelines in how to add CSS to be reused in several modules, so I'll create a draft of the general recommended CSS uses and a thread to discuss it.

But meanwhile, did you try Ivan's comment "using <aui:script require> to request the @liferay/document-library-preview-css" to see if it worked?

Reply:
@nhpatt @izaera it doesn't work, but actually curren change doesn't work either, I guess I tested it wrong before.

With current change (`link href`)

```
Refused to apply style from 'http://localhost:8080/o/document-library-preview-css/css/main.css?browserId=chrome&themeId=admin_WAR_admintheme&languageId=en_US&b=7401&t=1622474376000' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.
```

With `aui:require`
```

Reply:
In current examples we have in portal, sometimes we have  `type="text/css"` and some don't. Have you tried adding it back? Also, what's the status code returned if you type `http://localhost:8080/o/document-library-preview-css/css/main.css?browserId=chrome&themeId=admin_WAR_admintheme&languageId=en_US&b=7401&t=1622474376000` in the browser? I'm suspecting of a 404 due to some packaging issue and the .css not being generated. Just guessing...

Reply:
`http://localhost:8080/o/document-library-preview-css/css/main.css?browserId=chrome&themeId=admin_WAR_admintheme&languageId=en_US&b=7401&t=1622474376000` that takes me to the home page.

What I don't know is what I'm missing for `aui:require` not being working here

Reply:
Help me ~Obi Wan Kenobi~ @izaera you're my only hope

Reply:
`<aui:script require="@liferay/document-library-preview-css"></aui:script>` doesn't work because you are missing the version number and all that stuff... I think you need to add `document-library-preview-css` to the `package.json` file of `document-library-preview-audio`...

Reply:
`document-library-preview-css` is in the `package.json` of `document-library-preview-audio`. Do we really need to add the version? Is not possible for the loader to get the version defined there? (in this case is `*` but I guess that points to a certain version somewhere)

Reply:
>  to get the version defined there

"there" where?

Reply:
> "there" where?

Dunno, how the loader knows which versions needs to be used when you import a module from a React file, for example, when you define the version as `*` in the package.json?

Reply:
Ah OK, now I see what happens...

The loader gets the version from the `package.json` but when you use bundler-imports, the versions configured in the global `npmscripts.config.js` are directly injected into to the output `package.json`. That's why anything you bundler-import works even if you don't declare it in the `package.json`.

Also, when using a bundler-import the bundler rewrites all `require`s to prepend them with the provider name instead of the project's one. So, instead of turning `require('react')` into `require('my-project$react@1.0.0')`, it generates it as `require('my-provider$react@1.0.0')`.

Those two features makes everything work in the context of the loader.

But your `<aui:script require>` example is running outside of the loader context because it is called from a JSP, so it needs one step more, to be able to bootstrap the loader logic. That's the `NPMResolver` that reads the `package.json` in the OSGi bundle where the JSP lives, to find out the version number that is needed to bootstrap everything.

So, in your case, all the information the `NPMResolver` has is:

- `@liferay/document-library-preview-css` (because it's in the `<aui:script require>`).
- The `package.json` of `document-library-preview-audio` which will have been injected with bundler-imports.

But it won't have the information to know that `@liferay/document-library-preview-css` lives inside the `document-library-preview-css` OSGi bundle, so it will think it should be taken from `document-library-preview-audio` itself, thus the failure.

The only way to fix this error is by manually doing the work of the bundler and prefixing `@liferay/document-library-preview-css` with the name of the OSGi bundle, but we don't want to do that because it's too low level, complex, and error prone.

So I suggest using the old syntax to include the CSS in the JSP using its `/o/whatever/...` path and perhaps, if we really like the possibility of using a magic tag to include CSSs using the AMD loader, add a new task to the backlog to do it in a proper way...

Reply:
Thanks for the explanation @izaera (I'm saving it in my bookmark xD).

So... I have to leave it as it is right now. Could any of you test it and confirm that is working? (I've just added the steps to test it in the description).

If it works and all of you agree we can forward this one I guess.